### DISCUSSION TITLE
HELP!! flutter_ftms 0.6.2

### DISCUSSION THREAD

**Discussion by GustavoFonseca007** [community/community#56318] [Programming Help]
### Select Topic Area

Question

### Body

Hello, my friends. Could you help me? I'm using "flutter_ftms: ^0.6.2" and "flutter_blue: ^0.8.0. I try to use this code, but my Visual Studio isn't recognizing the package "flutter_ftms". There's a red line under "FlutterFtms" and "FlutterFtms" in the line "FlutterFtms ftms = FlutterFtms.instance;". There's another red line under "BluetoothDevice" in the line "List<BluetoothDevice> devices = [];", and a red line under "BluetoothDevice" in the line "BluetoothDevice? selectedDevice;". Also, there's a red line under "BluetoothDevice" in the line "final BluetoothDevice device;", and a red line under "FlutterFtms" and "FlutterFtms" in the line "FlutterFtms ftms = FlutterFtms.instance;", as well as a red line under "StreamSubscription" in the line "late StreamSubscription subscription;". Here's the code:

import 'package:flutter/material.dart';
import 'package:flutter_blue/flutter_blue.dart';
import 'package:flutter_ftms/flutter_ftms.dart';

class BluetoothPage extends StatefulWidget {
  @override
  _BluetoothPageState createState() => _BluetoothPageState();
}

class _BluetoothPageState extends State<BluetoothPage> {
  FlutterFtms ftms = FlutterFtms.instance;
  List<BluetoothDevice> devices = [];
  BluetoothDevice? selectedDevice;

  @override
  void initState() {
    super.initState();
    ftms.scanForFTMSDevices().listen((device) {
      setState(() {
        devices.add(device);
      });
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Bluetooth'),
      ),
      body: ListView.builder(
        itemCount: devices.length,
        itemBuilder: (context, index) {
          return ListTile(
            title: Text(devices[index].name),
            onTap: () async {
              selectedDevice = devices[index];
              await Navigator.push(
                context,
                MaterialPageRoute(
                  builder: (context) => MachinePage(selectedDevice!),
                ),
              );
            },
          );
        },
      ),
    );
  }
}

class MachinePage extends StatefulWidget {
  final BluetoothDevice device;

  MachinePage(this.device);

  @override
  _MachinePageState createState() => _MachinePageState();
}

class _MachinePageState extends State<MachinePage> {
  FlutterFtms ftms = FlutterFtms.instance;
  late StreamSubscription subscription;
  String display = '';

  @override
  void initState() {
    super.initState();
    subscription = ftms.connect(widget.device).listen((data) {
      setState(() {
        display = data.toString();
      });
    });
  }

  @override
  void dispose() {
    super.dispose();
    subscription.cancel();
    ftms.disconnect();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.device.name),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Text(display),
            Row(
              mainAxisAlignment: MainAxisAlignment.center,
              children: [
                ElevatedButton(
                  onPressed: () {},
                  child: Text('Stop'),
                ),
                SizedBox(width: 16),
                ElevatedButton(
                  onPressed: () {},
                  child: Text('Play'),
                ),
              ],
            ),
          ],
        ),
      ),
    );
  }
}

**Comment by Malte2036** [comment]
Why do you try `FlutterFtms ftms = FlutterFtms.instance`? You only have to follow the "Gettng Started" Guide of flutter_blue_plus (not the "Usage" Part). You do not need to install flutter_blue_plus by yourself. 

To use `flutter_ftms` follow the Documentation provided on the `flutter_ftms` [github](https://github.com/Malte2036/flutter_ftms) oder [pub.dev](https://pub.dev/packages/flutter_ftms#installation) page.

**Reply by GustavoFonseca007** [reply]
What do you think of this code? I haven't tested the code yet.


import 'package:flutter/material.dart';
import 'package:flutter_blue_plus/flutter_blue_plus.dart';
import 'package:flutter_ftms/flutter_ftms.dart';

class FtmsExample extends StatefulWidget {
 @override
 _FtmsExampleState createState() => _FtmsExampleState();
}

class _FtmsExampleState extends State<FtmsExample> {
 final FlutterBluePlus flutterBluePlus = FlutterBluePlus.instance;
 List<BluetoothDevice> devicesList = [];
 bool isSearching = false;
 String deviceData = '';

 @override
 Widget build(BuildContext context) {
 return Scaffold(
 appBar: AppBar(
 title: Text('FTMS Example'),
 centerTitle: true,
 ),
 body: Center(
 child: Column(
 mainAxisAlignment: MainAxisAlignment.center,
 children: [
 GestureDetector(
 child: Icon(Icons.bluetooth_searching, size: 100),
 onTap: () async {
 setState(() {
 isSearching = true;
 });
 await searchBluetoothDevices();
 setState(() {
 isSearching = false;
 });
 showModalBottomSheet(
 context: context,
 builder: (BuildContext context) {
 return Column(
 mainAxisSize: MainAxisSize.min,
 children: <Widget>[
 Padding(
 padding: EdgeInsets.only(top: 20),
 child: Text(
 'Dispositivos encontrados:',
 style: TextStyle(
 fontSize: 20,
 color: Colors.black,
 ),
 ),
 ),
 SizedBox(height: 20),
 Expanded(
 child: ListView.builder(
 shrinkWrap: true,
 itemCount: devicesList.length,
 itemBuilder: (BuildContext context, int index) {
 BluetoothDevice device = devicesList[index];
 return ListTile(
 title: Text(device.name ?? 'Unknown Device'),
 subtitle: Text(device.id.toString()),
 onTap: () async {
 await connectToDevice(device);
 },
 );
 },
 ),
 ),
 Padding(
 padding: EdgeInsets.only(bottom: 20),
 ),
 ],
 );
 },
 );
 },
 ),
 SizedBox(height: 20),
 Text(deviceData),
 SizedBox(height: 20),
 isSearching ? CircularProgressIndicator() : Container(),
 ],
 ),
 ),
 );
 }

 Future<void> searchBluetoothDevices() async {
 devicesList.clear();
 flutterBluePlus.startScan(timeout: Duration(seconds: 4));
 flutterBluePlus.scanResults.listen((results) {
 for (ScanResult result in results) {
 BluetoothDevice device = result.device;
 if (device.name.startsWith("DFIT")) {
 devicesList.add(device);
 }
 }
 });
 await flutterBluePlus.stopScan();
 for (BluetoothDevice device in devicesList) {
 await FTMS.connectToFTMSDevice(device);
 if (await FTMS.isBluetoothDeviceFTMSDevice(device)) {
 print('${device.name} is an FTMS device');
 } else {
 print('${device.name} is not an FTMS device');
 }
 }
 }

 Future<void> connectToDevice(BluetoothDevice device) async {
 await FTMS.connectToFTMSDevice(device);
 bool isFTMSDevice = await FTMS.isBluetoothDeviceFTMSDevice(device);
 if (isFTMSDevice) {
 print('${device.name} is an FTMS device');
 DeviceDataType? dataType = await FTMS.getDeviceDataType(device);
 if (dataType != null) {
 String deviceTypeString = FTMS.convertDeviceDataTypeToString(dataType);
 print('Device type: $deviceTypeString');
 }
 await FTMS.useDeviceDataCharacteristic(device, (DeviceData data) {
 // handle new data
 setState(() {
 deviceData = data.toString();
 });
 });
 MachineFeature? feature =
 await FTMS.readMachineFeatureCharacteristic(device);
 if (feature != null) {
 // handle feature object
 print('Received machine feature from device');
 }
 await FTMS.useMachineStatusCharacteristic(device, (MachineStatus status) {
 // handle new machine status
 print('Received new machine status from device');
 });
 } else {
 print('${device.name} is not an FTMS device');
 }
 }
}

**Comment by Malte2036** [comment]
I added an `flutter_ftms` example [here](https://github.com/Malte2036/flutter_ftms/tree/main/example).

**Reply by GustavoFonseca007** [reply]
@Malte2036 
“Hello, why aren’t the values for each feature displayed? For example, next to time, at the beginning, it should have 00:00.”

![Screenshot_20230526_121727_br com gustavo ftms](https://github.com/community/community/assets/100773950/50ed3f39-c5cd-4893-9b45-0c49cea4c175)

**Reply by Malte2036** [reply]
If you click on "use FTMS", you will get the FTMS device data. But you are right. Things like "energy" or "time" are missing. This is one of my next [tasks](https://github.com/Malte2036/flutter_ftms/issues/1) to fix this.

**Reply by GustavoFonseca007** [reply]
@Malte2036  Good evening, when I click on the button nothing appears.

**Reply by Malte2036** [reply]
Ok. Is there an flutter_ftms error in the console. Which device type is your fitness device?

**Reply by GustavoFonseca007** [reply]
@Malte2036  All I know is that the machine I tested is a ergometric rower that has ftms and bluetooth. I don't have the technical documentation for it, just the leaflet that teaches how to assemble it. The supplier (China) did not want to pass on any information about the machine.

**Reply by Malte2036** [reply]
I have not test the rower device type on an real machine. Could you provide the flutter log after you click "use FTMS"

**Reply by GustavoFonseca007** [reply]
@Malte2036 What machines have you tested it on?

**Reply by Malte2036** [reply]
I only tested it on an real cross trainer (and on an cross trainer, rower, treadmill and indoor bike virtual simulator).

**Reply by GustavoFonseca007** [reply]
@Malte2036 next monday i will test your app with the other machines and then i will tell you if it worked or not.
thx

**Reply by Malte2036** [reply]
Perfect. If it does not work,  send me the flutter log.

**Reply by Malte2036** [reply]
And open an issue directly on [https://github.com/Malte2036/flutter_ftms/issues](https://github.com/Malte2036/flutter_ftms/issues) then. thanks.
